### PR TITLE
feat: Add support for DataContext inheritance for non-UIElement

### PR DESCRIPTION
**Pull request by jeromelaban**
## PR Type

What kind of change does this PR introduce?
- Bugfix

## What is the new behavior?

The `DataContext` now propagates properly to non-`UIElement` `DependencyObject`-typed values in dependency properties. This behavior is generally used to data bind dependency objects in XAML for advanced configurations.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs [hunk 1]
```diff
@@ -561,43 +562,69 @@ private void OnDependencyPropertyChanged(DependencyPropertyDetails propertyDetai
...
-		{
-			if (
-				propertyDetails.Property == _templatedParentProperty
-				|| propertyDetails.Property == _dataContextProperty
+		void GetPropertyInheritanceConfiguration(
+			DependencyPropertyDetails propertyDetails,
+			DependencyPropertyChangedEventArgs args,
+			out bool hasValueInherits,
+			out bool hasValueDoesNotInherit,
+			out IDependencyObjectStoreProvider? valueAsProvider
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:604]: I'm not really sure why this `out` parameter is needed? Seems like the `as` could be done just as easily in the calling method.

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs [hunk 2]
```diff
@@ -561,43 +562,69 @@ private void OnDependencyPropertyChanged(DependencyPropertyDetails propertyDetai
...
+			GetPropertyInheritanceConfiguration(
+				propertyDetails,
+				args,
+				hasValueInherits: out var hasValueInherits,
+				hasValueDoesNotInherit: out var hasValueDoesNotInherits,
+				valueAsProvider: out var newValueAsProvider
+			);
 
-			if (!hasValueDoesNotInherit && hasValueInherits)
+			if (!hasValueDoesNotInherits)
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:573]: This block of conditions starts to get quite hard to follow... I wonder if it'd be more readable with a nested method like:

```csharp
SetChildrenBindable(object? obj) => _childrenBindable[GetOrCreateChildBindablePropertyIndex(propertyDetails.Property)] = obj;
```

**Review by davidjohnoliver — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs (hunk 1 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:604]: The reason for the out is to avoid casting twice, once inside the method, and another one outside if either of the two boolean not true, but also because it removes the use of a generic ValueTuple.

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:604]: Reconsidering it, multiple alterations actually don't need this anymore. Will remove.
